packages/ingest/granule.js

```diff
@@ -472,99 +468,63 @@ class Granule {
-
-    // compare the checksum of the existing file and new file, and handle them accordingly
-    if (renamingFile) {
-      const existingFileSum = await
-      aws.calculateS3ObjectChecksum({ algorithm: (checksumType || 'CKSUM'), bucket, key: destinationKey });
-
-      const stagedFileSum = checksumValue
-      || await aws.calculateS3ObjectChecksum({ algorithm: 'CKSUM', bucket, key: stagedFileKey });
-
-      // if the checksum of the existing file is the same as the new one, keep the existing file,
-      // else rename the existing file, and both files are part of the granule.
-      if (existingFileSum === stagedFileSum) {
-        await aws.deleteS3Object(bucket, stagedFileKey);
-      } else {
-        log.debug(`Renaming file to ${destinationKey}`);
-        await exports.renameS3FileWithTimestamp(bucket, destinationKey);
-        await exports.moveGranuleFile(
-          { Bucket: bucket, Key: stagedFileKey }, { Bucket: bucket, Key: destinationKey }
-        );
-      }
+    const s3ObjAlreadyExists = await aws.s3ObjectExists(
+      { Bucket: destinationBucket, Key: destinationKey }
+    );
+    log.debug(`file ${destinationKey} exists in ${destinationBucket}: ${s3ObjAlreadyExists}`);
+
+    let versionedFiles = [];
+    if (s3ObjAlreadyExists) {
+      stagedFile.duplicate_found = true;
+      const stagedFileKey = `${destinationKey}.${uuidv4()}`;
+      // returns renamed files for 'version', otherwise empty array
+      versionedFiles = await exports.handleDuplicateFile({
+        source: { Bucket: destinationBucket, Key: stagedFileKey },
+        target: { Bucket: destinationBucket, Key: destinationKey },
+        duplicateHandling,
+        checksumFunction: this.verifyFile.bind(this, file),
+        syncFileFunction
+      });
+    } else {
+      log.debug(`await sync file ${fileRemotePath} to s3://${destinationBucket}/${destinationKey}`);
+      await syncFileFunction(destinationBucket, destinationKey);
+      // Verify file integrity
+      log.debug(`await verifyFile ${JSON.stringify(file)}, s3://${destinationBucket}/${destinationKey}`);
+      await this.verifyFile(file, destinationBucket, destinationKey);
     }
 
-    const renamedFiles = (duplicateHandling === 'version')
-      ? await exports.getRenamedS3File(bucket, destinationKey) : [];
-
-    // return all files, the renamed files don't have the same properties(name, fileSize, checksum)
-    // from input file
-    return renamedFiles.concat({ Bucket: bucket, Key: destinationKey }).map((f) => {
-      if (f.Key === destinationKey) return stagedFile;
-      return {
+    // Set final filesize
+    stagedFile.fileSize = await aws.getObjectSize(destinationBucket, destinationKey);
+    // return all files, the renamed files don't have the same properties
+    // (name, fileSize, checksum) as input file
+    log.debug(`returning ${JSON.stringify(stagedFile)}`);
+    return [stagedFile].concat(versionedFiles.map((f) => (
```

Comment:
It's important to note the return value from handleDuplicateFiles, which for all cases other than `version` is `[]`. This means that `[stagedFile].concat()` will concatenate an empty array, so it will return an array with only the staged file. Only for the `version` case does it return all versioned filenames.

Reply:
There is probably an underlying assumption that the duplicateHandling will not change once a granule has been ingested here, and that may be a wrong assumption, but in the interest of only refactoring to simplify, I have maintained current inputs & outputs. If you think we should address that by, say, returning any versioned filenames for a `replace` granule, I'm not inclined to disagree with you, but I do think that would be out of scope for this ticket/PR.